src/cpp/rtps/writer/StatefulWriter.cpp

```diff
@@ -768,12 +768,12 @@ void StatefulWriter::check_acked_status()
         if(mp_listener != nullptr)
         {
             std::vector<CacheChange_t*> all_acked_changes;
-            for(SequenceNumber_t current_seq = get_seq_num_min(); current_seq <= min_low_mark; ++current_seq)
             {
-                for(std::vector<CacheChange_t*>::iterator cit = mp_history->changesBegin();
-                        cit != mp_history->changesEnd(); ++cit)
+                std::lock_guard<std::recursive_mutex> guard(*mp_history->getMutex());
+                SequenceNumber_t min_sequence = get_seq_num_min();
+                for (auto cit = mp_history->changesBegin(); cit != mp_history->changesEnd(); ++cit)
                 {
-                    if((*cit)->sequenceNumber == current_seq)
+                    if ((*cit)->sequenceNumber >= min_sequence || (*cit)->sequenceNumber <= min_low_mark)
```

Comment:
Cannot be a `||`. It should be an `&&`. Another optimization suggested by @MiguelCompany: 

```suggestion
                    // History is order by sequence number.
                    // We can break if the analyzed CacheChange is already greater than 'min_low_mark'
                    if ((*cit)->sequenceNumber > min_low_mark)
                    {
                        break;
                    }
                    else
```

Reply:
Apologies as I had reversed the logic but forgot the operator.  If the history is sorted by sequence, then the suggested change is great.

Reply:
> If the history is sorted by sequence, then the suggested change is great.

Yes, the history is sorted by sequence. Please apply the suggestions so the CI will pass and this could be merged.